File: Vowel Balance/Vowel Balance.py
# Checks if the number of vowels in the first half and second half of a string are equal.
# Returns True if balanced, False otherwise.
# Parameters:
#   s: input string
# Returns:
#   True if the string is vowel balanced, False otherwise.
#
# Usage Examples:
#   is_balanced("racecar") => True
#   is_balanced("hello")   => False
# Time Complexity: O(n), where n is the length of the input string
# Space Complexity: O(1), only a few variables are used
def is_balanced(s: str) -> bool:
    vowels = set("aeiouAEIOU")
    n = len(s)
    
    # If odd, ignore the middle char
    half = n // 2
    first_half = s[:half]
    second_half = s[-half:] if half > 0 else ""
    
    count_first = sum(1 for ch in first_half if ch in vowels)
    count_second = sum(1 for ch in second_half if ch in vowels)
    
    return count_first == count_second

File: Vowel Balance/Vowel Balance.rs
// Time Complexity: O(n), where n is the length of the input string
// Space Complexity: O(1), only a few variables are used
fn is_balanced(s: &str) -> bool {
    let vowels = "aeiouAEIOU";
    let chars: Vec<char> = s.chars().collect();
    let n = chars.len();
    let half = n / 2;

    let count_first = chars[..half].iter().filter(|c| vowels.contains(*c)).count();
    let count_second = chars[n - half..].iter().filter(|c| vowels.contains(*c)).count();

    count_first == count_second
}

fn main() {
    println!("{}", is_balanced("racecar")); // true
    println!("{}", is_balanced("Lorem Ipsum")); // true
    println!("{}", is_balanced("Kitty Ipsum")); // false
}
